ykrt/src/compile/jitc_yk/gdb.rs

```diff
@@ -0,0 +1,241 @@
+//! This implements support for the GDB JIT Compilation Interface described here:
+//! https://sourceware.org/gdb/onlinedocs/gdb/JIT-Interface.html
+//!
+//! This allows gdb to recognise our JITted code, so that we can have higher-level information
+//! (than just raw asm) displayed when debugging traces.
+
+use deku::prelude::*;
+use indexmap::IndexMap;
+use std::{
+    ffi::{c_char, c_int, CString},
+    io::Write,
+    ptr,
+    sync::Mutex,
+};
+use tempfile::NamedTempFile;
+
+/// JITted code actions.
+///
+/// This is a mirror of `jit_actions_t` from <jit-reader.h>.
+#[allow(dead_code)]
+#[repr(u32)]
+pub enum JitActionsT {
+    JitNoAction = 0u32,
+    JitRegisterFn = 1u32,
+    JitUnregisterFn = 2u32,
+}
+
+/// An entry in gdb's linked list of JITted code objects.
+///
+/// This is a mirror of `struct jit_code_entry` from <jit-reader.h>.
+#[repr(C)]
+#[derive(Debug)]
+pub struct JitCodeEntry {
+    next_entry: *mut Self,
+    prev_entry: *mut Self,
+    symfile_addr: *const c_char,
+    symfile_size: u64,
+}
+
+impl Drop for JitCodeEntry {
+    fn drop(&mut self) {
+        let lock = LOCK.lock();
+        let prev: *mut Self = self.prev_entry;
+        let next: *mut Self = self.next_entry;
+
+        // Take the entry out of gdb's linked list.
+        //
+        // If there's a previous entry, update its "next" link.
+        if !prev.is_null() {
+            unsafe { ptr::write(ptr::addr_of_mut!((*prev).next_entry), next) };
+        } else {
+            // The entry being deleted is the head of the list. Update the head pointer.
+            unsafe { __jit_debug_descriptor.first_entry = next };
+        }
+        // If there's a "next" entry, update its "prev" link.
+        if !next.is_null() {
+            unsafe { ptr::write(ptr::addr_of_mut!((*next).prev_entry), prev) };
+        }
+        drop(lock);
+    }
+}
+
+/// The top-level data structure for talking to gdb.
+///
+/// This is a mirror of `struct jit_descriptor` from <jit-reader.h>.
+#[repr(C)]
+pub struct JitDescriptor {
+    version: u32,
+    action_flag: u32,
+    relevant_entry: *mut JitCodeEntry,
+    first_entry: *mut JitCodeEntry,
+}
+
+/// An instance of [JitDescriptor] with a special symbol name that gdb recognises specially.
+#[allow(non_upper_case_globals)]
+#[no_mangle]
+pub static mut __jit_debug_descriptor: JitDescriptor = JitDescriptor {
+    version: 1,
+    action_flag: 0,
+    relevant_entry: ptr::null_mut(),
+    first_entry: ptr::null_mut(),
+};
+
+/// Due to the way the gdb JIT API works, we are unable to have Rust handle the synchronisation of
+/// `__jit_debug_descriptor` for us.
+unsafe impl Send for JitDescriptor {}
+unsafe impl Sync for JitDescriptor {}
+unsafe impl Send for JitCodeEntry {}
+unsafe impl Sync for JitCodeEntry {}
+
+/// Guards `__jit_debug_descriptor`.
+static LOCK: Mutex<()> = Mutex::new(());
```

Comment:
You might not be able to wrap `__jit_debug_descriptor` inside a `Mutex` because gdb expects a certain layout, but you can tie it to a Mutex using a `with`-style API. I think this will be better because you can: a) remove a lot of the `unsafe` blocks associated with accessing the global static; and b) ensure you never forget to lock and unlock accesses to `__jit_debug_descriptor`. I've mocked up a little example of what I mean here:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=57000cb92db0673bf5f4d9ba398b3ba3

Reply:
I remember discussing this with you a few months back. Seems like a good idea, but I may defer it for a second PR if that's OK?

Reply:
I also had the same worry Jake has: I don't think the `Mutex` is guaranteed to lock what you want it to lock. I didn't have a good solution, though. Jake's seems easy-ish to implement?

Reply:
OK, I'll do it from the get-go then.